src/main/java/com/AjaxController.java

```diff
@@ -1,107 +1,158 @@
+    @Autowired
+    private PerkService perkService;
+    @Autowired
+    private SubscriptionService subscriptionService;
+
+    User user;
+
+    /**
+    public static final Comparator<Perk> DESCENDING_COMPARATOR = new Comparator<Perk>() {
+        // Overriding the compare method to sort the age
+        public int compare(Perk p, Perk p1) {
+            return p.getScore()-p1.getScore();
+        }
+    };
+    **/
+
+    /*
+     * Get table rows for a user populated with that users subscription and perk names.
+     * @Param: userName=userName of user
+     * @Ret:HTML rows of that users subscription and perks
+     * */
+    @ResponseStatus(HttpStatus.OK)
+    @RequestMapping(method=RequestMethod.GET, value = "/GetTable")
+    public String GetTable(@RequestParam String userName)
+    {
+        String ret = "";
+        user = userService.findByUsername(userName);//Find user
+        List<Subscription> subs = user.getSubscriptions();
+
+        /* For all subscriptions get all perks and append table row. */
+        for(int i = 0; i < subs.size(); i++) {
+            for(int n = 0; n < subs.get(i).getPerks().size(); n++) {
+                ret += "<tr><td id = \"subscription_name\">" + subs.get(i).getName() + "</td>" + "<td id = \"perk_name\">" + subs.get(i).getPerks().get(n).getName() + "</td>" +
+                        "<td><button class=\"upvotebutton\">Upvote</button></td>" +
+                        "<td><button class=\"downvotebutton\">Downvote</button></td>" +
+                        "<td id = \"score_id\">" + subs.get(i).getPerks().get(n).getScore() + "</td></tr>";
+            }
+        }
+
+        return ret;
+    }
+
+
+    /*
+     * Add subscription to user.
+     * @Param: userName=userName of user
+     * @Param: perkJson=Json of the subscription to be added
+     * */
+    @ResponseStatus(HttpStatus.OK)
+    @RequestMapping(method=RequestMethod.POST, value = "/AddSubscription",consumes = {"application/json"})
+    public void AddSubscription(@RequestParam String userName,@RequestBody Subscription subscriptionJson)
+    {
+        user = userService.findByUsername(userName);
+
+        if (!subscriptionService.existsByName(subscriptionJson.getName())) {
+            subscriptionService.save(subscriptionJson);
+        }
+
+        Subscription newSub = subscriptionService.findByName(subscriptionJson.getName());
+        if (!user.getSubscriptions().contains(newSub)) {
+            user.addSubscription(newSub);
+            userService.save(user);
+        }
+    }
+
+
+    /*
+     * Add perk to a users subscription.
+     * @Param: userName=userName of user
+     * @Param: subName=name of subscription
+     * @Param: perkJson=Json of the perk to be added
+     * */
+    @ResponseStatus(HttpStatus.OK)
+    @RequestMapping(method=RequestMethod.POST, value = "/AddPerk",consumes = {"application/json"})
+    public void AddPerk(@RequestParam String userName,@RequestParam String subName,@RequestBody Perk perkJson) {

```

Comment:
It looks like you have an old AddPerk method. Can you change it back?

Reply:
Done.